{% extends "base.html" %}
{% load static i18n %}

{% block includes %}
<script type="text/javascript">
$(function() {
  $('#help_navbar_button').addClass('active');      
});
</script>
{% endblock %}

{% block content %}
<div class="container" style="margin-top:15px;">
  <div class="row">
    <div class="col-xs-2">
      <ul class="nav nav-pills nav-stacked">
        <li role="presentation" class="active"><a href="#background" aria-controls="background" role="tab" data-toggle="tab">Background</a></li>
        <li role="presentation"><a href="#nomenclature" aria-controls="nomencalture" role="tab" data-toggle="tab">Nomenclature</a></li>
        <li role="presentation"><a href="#classification" aria-controls="classification" role="tab" data-toggle="tab">Classification</a></li>
        <li role="presentation"><a href="#download" aria-controls="download" role="tab" data-toggle="tab">Download</a></li>
        <li role="presentation"><a href="#cite" aria-controls="cite" role="tab" data-toggle="tab">How To Cite</a></li>
        <li role="presentation"><a href="#contact" aria-controls="contact" role="tab" data-toggle="tab">Contact</a></li>

      </ul><br />
{#      <div class="center-block">#}
{#        <a class="github-button" href="https://github.com/edraizen/HistoneDB" data-style="mega" data-count-href="/edraizen/HistoneDB/watchers" data-count-api="/repos/edraizen/HistoneDB#subscribers_count" data-count-aria-label="# watchers on GitHub" aria-label="Watch edraizen/HistoneDB on GitHub" style="width:100%;">Watch</a><br />#}
{#        <a class="github-button" href="https://github.com/edraizen/HistoneDB/issues" data-icon="octicon-issue-opened" data-style="mega" data-count-api="/repos/edraizen/HistoneDB#open_issues_count" data-count-aria-label="# issues on GitHub" aria-label="Issue edraizen/HistoneDB on GitHub">Issue  </a>#}
{#      </div>#}
    </div>
    <div class="col-xs-10">
      <div class="tab-content">
        <div role="tabpanel" class="tab-pane active" id="background">
          <h2 id="background">Background information</h2>
            <h3 id="background">Nucleosome structure</h3>
          Nucleosomes are the basic units of chromatin compaction. Nucleosome core particle (NCP) wraps around 147 bp of DNA in ~1.7 left-handed super helical turns around
an octamer of four types of core hisotnes (H3,
H4, H2A and H2B — two copies of each) [1,2]. NCPs are further connected by linker DNA strands and a linker histone H1 further interacts with the NCPs and linker DNA to facilitate formation of higher order chromatin structures.
            Nucleosomes play key roles in epigenetic regulation of transcription,
replication, DNA repair and cell reprogramming.<br><br>
          <div class="row">
          <div class="col-xs-6">
            <img  src="{% static 'browse/img/assembly.png' %}" width="100%" />
          </div>
          <div class="col-xs-6">
                    <video controls="controls"  width="400px" height="300px" poster="{% static 'browse/video/poster.jpg' %}">
                    <source src="{% static 'browse/video/mdnucl.mp4' %}" type="video/mp4">
                    <source src="{% static 'browse/video/mdnucl.webm' %}" type="video/webm">
                    </video>
          </div>
          </div>
          <div class="row">
              <div class="col-xs-6">Nucleosome organization diagram (adapted from [3-4]).</div>

              <div class="col-xs-6">Dynamic representation of the nucleosome with linker DNA [4].</div>
          </div>
             <h3 id="background">Histone variants</h3>
            Most eukaryotes have histone isoforms
            for some or all of the histone families (H2A, H2B, H3, H4, H1) [5].
            Classification of histone isoforms is a daunting task. They are usually subdivided into ‘canonical’ replication-dependent histones that are expressed during the S-phase of cell cycle and replication-independent histone ‘variants’,
            constitutively expressed during cell cycle. In animals, genes encoding canonical histones are typically clustered along the chromosome, lack introns, and employ a specific type of regulation at the RNA level with a stem loop structure at the 3’ end instead of polyA tail.
            On the other hand, genes encoding histone variants are usually not clustered, have introns, and their mRNAs are regulated with polyA tails similar to the mRNAs of most genes.
            In plants, canonical histone genes lack introns, but are not clustered and the mRNAs are polyadenylated. Remarkably, more complex multicellular organisms typically have a higher number of histone variants providing a variety of different functions.
            Recent data are accumulating about the roles of diverse histone variants highlighting the functional links between variants and the delicate regulation of organism development.

    <h3 id="background">References</h3>

            <ol>

                        <li>Kornberg RD. "Chromatin structure: a repeating unit of histones and DNA." <i>Science</i>, 1974. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=4825889">4825889</a></li>

                        <li>Luger K, Mader AW, et al. "Crystal structure of the nucleosome core particle at 2.8 A resolution." <i>Nature</i>, 1997. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=9305837">9305837</a></li>

                        <li>Shaytan AK, Landsman D, et al. "Nucleosome adaptability conferred by sequence and structural variations in histone H2A-H2B dimers." <i>Curr Opin Struct Biol</i>, 2015. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=25731851">25731851</a></li>

                        <li>Shaytan AK, Armeev GA, et al. "Coupling between histone conformations and DNA geometry in nucleosomes on a microsecond timescale: atomistic insights into nucleosome functions." <i>J Mol. Biol.</i>, 2015. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=26699921">26699921</a></li>

                        <li>Talbert PB, Ahmad K, et al. "A unified phylogeny-based nomenclature for histone variants." <i>Epigenetics Chromatin</i>, 2012. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=22650316">22650316</a></li>
            </ol>

        </div>
        <div role="tabpanel" class="tab-pane" id="contact">
          <h2 id="contact">Contact</h2>
          <div class="container">
            <div class="row">
              <div class="col-xs-12">
                  <!-- Place this tag where you want the button to render. -->
                <!-- Place this tag where you want the button to render. -->
{#                <a class="github-button" href="https://github.com/edraizen/HistoneDB" data-style="mega" data-count-href="/edraizen/HistoneDB/watchers" data-count-api="/repos/edraizen/HistoneDB#subscribers_count" data-count-aria-label="# watchers on GitHub" aria-label="Watch edraizen/HistoneDB on GitHub">Watch</a>#}
{#                <a class="github-button" href="https://github.com/edraizen/HistoneDB/issues" data-icon="octicon-issue-opened" data-style="mega" data-count-api="/repos/edraizen/HistoneDB#open_issues_count" data-count-aria-label="# issues on GitHub" aria-label="Issue edraizen/HistoneDB on GitHub">Issue</a>#}
              </div>
            </div>
            <div class="row">
                Please, send your question and comments to <a href="mailto:histonedb-help@ncbi.nlm.nih.gov">histonedb-help@ncbi.nlm.nih.gov</a>.
{#              <div class="col-xs-4 center-block">#}
{#                <h3><a href="http://www.ncbi.nlm.nih.gov/CBBresearch/Panchenko/">Anna Panchenko</a></h3><br />#}
{#                Computational Biophysics Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                email here<br />#}
{#              </div>#}
{#              <div class="col-xs-4 center-block">#}
{#                <h3>Eli Draizen</h3><br />#}
{#                Computational Biophysics Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                eli.draizen _at_ nih.gov<br />#}
{#              </div>#}
{#              <div class="col-xs-4 center-block">#}
{#                <h3>Alexey K. Shaytan</h3><br />#}
{#                Computational Biophysics Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                alexey.shaytan _at_ nih.gov<br />#}
{#              </div>#}
            </div>
{#            <div class="row">#}
{#              <div class="col-xs-4 center-block">#}
{#                <h3><a href="http://www.ncbi.nlm.nih.gov/research/staff/landsman/">David Landsman</a></h3><br />#}
{#                Bioinformatics of Chromatin Structure Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                email here<br />#}
{#              </div>#}
{#              <div class="col-xs-4 center-block">#}
{#                <h3><a href="http://www.ncbi.nlm.nih.gov/research/staff/marino/">Leonardo Mariño-Ramírez</a></h3><br />#}
{#                Bioinformatics of Chromatin Structure Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                email here<br />#}
{#              </div>#}
{#          </div>#}
        </div>
      </div>
      <div role="tabpanel" class="tab-pane" id="nomenclature">
        <h2 id="nomencalture">Nomenclature</h2>
        This database of histone variant follows the new nomenclature introduced in ref. [1], which may be summarized in the following form: <br /><br />
        [prefix]<font color="red">TYPE</font>.<font color="blue">VARIANT</font><br />
        <ol>
          <li><font color="red">TYPE</font> = Histone type name (H2A, H2B, H3, H4, H1)</li>
          <li><font color="blue">VARIANT</font>:
            <ul>
              <li>Letter: used to denote a structurally distinct monophyletic clade of a histone family (exception is H2A.X).</li>
              <li>Number: used to denote a species-specific variant, not defined by phylogeny but clear orthology.</li>
            </ul>
          </li>
        </ol>
        prefix = used to denote a structurally distinct clade of core histones (e.g.: <b>macro</b>H2A, <b>sub</b>H2B, <b>cen</b>H3).<br />

          <div class="row">
               <div class="col-md-8">
        Example: <font color="red">H2A</font>.<font color="blue">Z</font>
</div>
               </div>
          <h3 id="background">References</h3>

            <ol>

                        <li>Talbert PB, Ahmad K, et al. "A unified phylogeny-based nomenclature for histone variants." <i>Epigenetics Chromatin</i>, 2012. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=22650316">22650316</a></li>
            </ol>

      </div>
      <div role="tabpanel" class="tab-pane" id="classification">
           <h2 id="nomencalture">Classification</h2>
          The automatic classification algorithms employed in the current database is described below. More details available in ref.[1]. <br><br>

       Curated alignments of histone variants were used to train Hidden Markov Models, utilizing HMMER 3.1b2,
          to create one HMM for each variant. These models were used as a part of automatic extraction and annotation
          pipeline. Namely, all sequences from the nr protein database have been classified by the HMM models
          (variants and canonical) and were added into the HistoneDB 2.0 as “automatically extracted sequences”.
          We assigned a model with a maximum HMMER score to a given sequence. The score was required to exceed
          a certain threshold identified as follows. For any given variant model, we searched the curated sequence
          set and calculated HMMER scores for all curated sequences. We then varied the HMMER score threshold
          in order to distinguish variant sequences from all other sequences with 90% specificity. Specificity
          of retrieval was estimated based on the number of true positives (TPs; sequences correctly predicted by
          their native model) and false positives (FPs; incorrectly predicted sequences) found above each HMMER
          score cutoff. The specificity was calculated as TN/(FP+TN). Then a desired score cutoff value was
          obtained from the interpolated inverse curve of the score cutoffs plotted versus specificity. The
          efficacy of classification based on HMM models depends significantly on whether the respective
          histone variant model is sufficiently divergent and forms a separate clade on the phylogenetic
          tree. For certain variants,
          which emerged repeatedly in the course of evolution, the classification becomes difficult unless a
          characteristic signature is known. In the case of H2A.X, HMM search algorithm was supplemented by the
          pattern matching, since this variant is characterized by the presence of ‘SQ(E/D)Φ’ motif as described
          above. Any sequence that was classified as canonical H2A by HMMER-based algorithm was reclassified as
          H2A.X if it had this motif at its C-terminus. The described above classification algorithm can be also
          applied to classify any sequence of interest via "Analyze Your Sequence" tab.

                   <h3 id="background">References</h3>

            <ol>

                        <li>
                            Draizen EJ, Shaytan AK, Marino-Ramirez L, Talbert PB, Landsman D, Panchenko AR. "HistoneDB 2.0 - with Variants: an integrated resource to explore histones and histone variants." <i>Database</i>. In Preparation.

                            </li>
            </ol>

        </div>
      <div role="tabpanel" class="tab-pane" id="download">
          <div class="container">
            <div class="row">
              <h2>Download</h2>
            </div>
            <div class="row">
              <h3>How can I download all sequences?</h3>
            </div>
            <div class="row">
              List of all histone sequences referenced by GIs can be downloaded <a href="{% static 'browse/dumps/seqs.txt' %}">here.</a><br>
              List of all HMM scores for the sequences against matching HMM models can be downloaded <a href="{% static 'browse/dumps/scores.txt' %}">here.</a><br>
            </div>
          </div>
      </div>

      <div role="tabpanel" class="tab-pane" id="cite">
          <h2 id="contact">How to cite</h2>
          Please cite the database as:<br>

        E.J. Draizen, A.K. Shaytan, L. Marino-Ramirez, P.B. Talbert, D. Landsman, A.R. Panchenko. "HistoneDB 2.0 - with Variants: an integrated resource to explore histones and histone variants." <i>Database</i>. In Preparation.
      </div>
    </div>
  </div>
</div>
<!-- Place this tag right after the last button or just before your close body tag. -->
{#<script async defer id="github-bjs" src="https://buttons.github.io/buttons.js"></script>#}
{% endblock %}
